### ISSUE TITLE
Expose container failures as events

### ISSUE THREAD

**Issue by rrati** [kubernetes/kubernetes#5927]
It appears the the status reflected in kubernetes via 'kubectl get pods' isn't always correct.  I'm seeing pods start on the minions then the app container (nginx in this case) dies while the pause container stays running.  The app container is continually restarted and continually dies 1-2 seconds after it starts.  The status for this pod in kubernetes is Running, but clearly that is not the case.  The information stored for one such pod in etcd looks like this:

```
{"kind":"Pod","id":"my-hostname-density100-038b4df2-d301-11e4-806b-54ee7510755kjfpx","uid":"03978f40-d301-11e4-a6fe-54ee7510755a","creationTimestamp":"2015-03-25T15:10:03Z","selfLink":"/api/v1beta1/pods/my-hostname-density100-038b4df2-d301-11e4-806b-54ee7510755kjfpx?namespace=e2e-density100-038b4df2-d301-11e4-806b-54ee7510755a","resourceVersion":51000,"apiVersion":"v1beta1","namespace":"e2e-density100-038b4df2-d301-11e4-806b-54ee7510755a","generateName":"my-hostname-density100-038b4df2-d301-11e4-806b-54ee7510755a-","labels":{"name":"my-hostname-density100-038b4df2-d301-11e4-806b-54ee7510755a"},"desiredState":{"manifest":{"version":"v1beta2","id":"","volumes":null,"containers":[{"name":"my-hostname-density100-038b4df2-d301-11e4-806b-54ee7510755a","image":"dockerfile/nginx","ports":[{"containerPort":80,"protocol":"TCP"}],"resources":{},"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"PullIfNotPresent","capabilities":{}}],"restartPolicy":{"always":{}},"dnsPolicy":"ClusterFirst"},"host":"127.0.0.1"},"currentState":{"manifest":{"version":"","id":"","volumes":null,"containers":null,"restartPolicy":{}},"status":"Running","Condition":[{"kind":"Ready","status":"None"}],"host":"127.0.0.1","podIP":"172.17.3.27","info":{"my-hostname-density100-038b4df2-d301-11e4-806b-54ee7510755a":{"state":{"termination":{"exitCode":127,"startedAt":"2015-03-25T15:14:28Z","finishedAt":"2015-03-25T15:14:28Z"}},"ready":false,"restartCount":7,"image":"dockerfile/nginx","imageID":"docker://9f4cd86c1aece2a4f468a3c39db6b13016457586f92abbc066dccccf42f63131","containerID":"docker://36098413f6174486405310e8db1561fd74f656f5282e4542cd8df1325c0573a9"}}}}
```

When it restarts the containerID changes, but the currentState:status never changes after it becomes Running.

A pod that has not yet started looks like this:

```
{"kind":"Pod","id":"my-hostname-density100-038b4df2-d301-11e4-806b-54ee7510755zs8zc","uid":"03903057-d301-11e4-a6fe-54ee7510755a","creationTimestamp":"2015-03-25T15:10:03Z","resourceVersion":49324,"apiVersion":"v1beta1","namespace":"e2e-density100-038b4df2-d301-11e4-806b-54ee7510755a","generateName":"my-hostname-density100-038b4df2-d301-11e4-806b-54ee7510755a-","labels":{"name":"my-hostname-density100-038b4df2-d301-11e4-806b-54ee7510755a"},"desiredState":{"manifest":{"version":"v1beta2","id":"","volumes":null,"containers":[{"name":"my-hostname-density100-038b4df2-d301-11e4-806b-54ee7510755a","image":"dockerfile/nginx","ports":[{"containerPort":80,"protocol":"TCP"}],"resources":{},"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"PullIfNotPresent","capabilities":{}}],"restartPolicy":{"always":{}},"dnsPolicy":"ClusterFirst"},"host":"127.0.0.1"},"currentState":{"manifest":{"version":"","id":"","volumes":null,"containers":null,"restartPolicy":{}},"status":"Waiting","host":"127.0.0.1"}}
```

It is lacking the currentState field..

I would expect that the kubelet should wait to verify that the container runs for a short time period before determining that it is running and thus report it back to the master.  I would also expect that the kubelet should report to the master when the pod isn't fully functional (ie a container dies and/or won't stay up after restart).

**Comment by yujuhong** [comment]
I think this is because kubelet sets the pod status right after syncing the containers. The syncing process'd restart the containers if necessary, and then report the good "running" status after that. Maybe we should track the crashing pods and report it back somehow?

**Comment by rrati** [comment]
I think the information needed to take action is already in the kubelet.  It A) knows that a containers state is terminated, and it knows that the docker container id is.  if the state is terminated, the pod can't be running.   If the docker container id changes something hiccuped and should be monitored for a while.  Maybe even increment a restart counter.  This information seems to be getting updated in etcd, so I think the kubelet might just need to perform some status checks and set pod status appropriately when sending updates.

That might bring up a whole different issue of what to do if a pod transitions from running->pending though.  imo I'd like to see more information about state transitions some place.  Why a pod changed state, was killed, etc.

**Comment by vmarmol** [comment]
It does sound like we need better support for identifying and handling frequent crashes.

At the very least I see the need to surface container failures in events. We show containers starts today, but failures can more clearly identify problems. Longer term, we probably need to identify frequent crashes and either evict the pod or change its status to something that reflects the issue (eviction is probably more correct).

As @yujuhong mentioned we only catch the errors when we do restart the containers so as far as we know, everything is okay when we report the event.

@rrati I think we should do the event reporting first and then file a lower-priority issue for better handling of crash-looping detection. There is some subtlety there and it's unclear how we will handle that going forward. WDYT?

**Comment by rrati** [comment]
@vmarmol I think the most important part of this is ensuring the state reflected in the master is the actual state of the pod.  If exposing the container failures in the events is intended to cause the state in the master to reflect that the pod is not running, errored, or anything other than running, then I'm fine making that the important first change and logging a separate PR for possible mitigation solutions.

How to handle the failure cases is definitely a more complicated issue.  I think we'll need to come up with some kind of basic policy and procedures in the short term so we can act upon the failure data that would exist in the events.

**Comment by yujuhong** [comment]
Agreed that surfacing the errors is a good first step.

Kubelet already records the restartCount for the container (which is currently capped at 5 because of garbage collection).  https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/types.go#L604
We can potentially leverage this for detecting crash loops.

One possible way to handle the crashes is to set the pod status to failed (which may disrespect the user-specified "always restart" policy, but I think it's acceptable). If the pod is associated with a replication controller, the RC can spin off a new pod. This *may* resolve the problem if the crashes were node-specific (in the hope that the new pod would be scheduled to another node). However, in the case where the pod itself is problematic and cannot be stably run on any node, RC would just enter a create-delete loop until the user notices the issue.

**Comment by vmarmol** [comment]
Filed #6014 for the second part of this problem. Will re-purpose this issue to expose the crashes as events. Thanks @rrati!

**Comment by vmarmol** [comment]
(arguably I should have done that the other way around, sorry about that...)

**Comment by dchen1107** [comment]
We already export container failure as events, and you can retrieve them through `kubectl get events` or `kubectl describe pods <pod-id>`

There is a bug #5957 related to retrieve pods related events filed by @vishh yesterday, I fixed it through #5994

**Comment by dchen1107** [comment]
I am going to close this one for now, and address the root issue #6014 or  #5668

**Comment by dchen1107** [comment]
@vmarmol is right, kubelet only reports restart events, no failure events. Reopen this one, but I don't think it is blocker for v1.0.

**Comment by rrati** [comment]
@dchen1107 Why wouldn't this be a blocker?  The core problem, imo, is that the state in the master isn't the actual state of the system.  If the master says a pod is running but that pod isn't actually running or is crashing, that's a big deal.  Are you saying the state issue will be resolved for 1.0 but not necessarily by exposing container failures in events?

**Comment by bgrant0607** [comment]
You might be interested in #7843
